### ISSUE TITLE
The VQ4P apply is NEON-only, so §47's tuning does not travel to x86

### ISSUE THREAD

**Issue by marcobambini** [sqliteai/warp#32]
`src/model.c:2137` gates the vectorized VQ4P apply:

```c
#if (defined(__ARM_NEON) || defined(__aarch64__)) && !defined(WASTE_P6_SCALAR)
```

The vectorized side is four `vqtbl4q_s8` per 16-row tile — a 64-byte table
lookup — over a `vld3q_u8` deinterleave, with widening adds into int16. An x86
build takes the `else` branch: one row at a time, four scalar table loads per
row. It is not a narrower kernel, it is the absence of one — and on the same
build VQ3R's table gets `simd_avx512.c`.

Reported by @fab2s in #11, while trying to run the 64-entry crossover
`docs/LEARNED.md` §50 left open. Confirmed here by reading rather than by
measurement: this repo has no x86 machine that is not emulated, which is also
why `docs/BACKENDS.md` records AVX-512 as "compiled and dispatched, never
executed".

## What is being assumed because of it

- **The 64-entry crossover cannot be measured on an x86 CPU at all.** §50 left
  open whether VQ4P's 72 KB int8 table inverts the LUT-versus-reconstruct
  result it measured at 256 entries. On x86 that comparison puts a scalar apply
  against a vectorized one, so the number would be answering an ISA question
  and not a table-size one.
- **§47's tuning table is an ARM result.** `WASTE_XPAR`, `WASTE_XPAR_BATCH`,
  `WASTE_P6_CHUNK` and the thread counts were measured against the vectorized
  apply, and `CLAUDE.md` carries them as engine-wide guidance. On x86 the kernel
  underneath them is a different kernel, and the inversion between Kimi-Linear
  and K3 has never been retested there.
- **VQ4P is the format `convert.py` produces by default**, so this is the path
  on every x86 build, not a corner case.

## What might close it

`vqtbl4q_s8` looks up 64 bytes across four registers. The AVX-512 analogue is
**VBMI's `vpermi2b`**, which indexes 128 bytes across two zmm registers — the
same primitive with room to spare for a 64-entry table. Zen 4 and later, and
Ice Lake and later, have VBMI. `src/simd_avx512.c` already exists as the place
it would go, and dispatch resolves once at init, so this is a slot to fill
rather than an `#ifdef` in `model.c`.

Unmeasured in both directions, and worth stating before anyone spends a
weekend on it: the path may still be bound by the deinterleave rather than by
the lookup, in which case the table primitive buys much less than it looks
like it should. Cross-lane byte permutes are also not uniformly cheap across
x86 microarchitectures. This is a hypothesis with an obvious first experiment,
not a plan.

## The doc sentence to qualify either way

`docs/BACKENDS.md` says:

> The third hot path, the VQ gather, gets nothing — no x86 SIMD helps it
> either, for the same reason NEON does not.

That was written about VQ3R's 256-entry table, where it holds: 256 entries fit
no byte-permute primitive. It is now being read as covering VQ4P, where the
premise it rests on is false. The sentence needs qualifying whether or not
anyone writes the kernel.

## Acceptance, if someone does write it

- **Bit-identity, not "close".** `-DWASTE_P6_SCALAR` already makes the ARM path
  self-checking against the portable one, and §43 says why the bar is that
  high: an int8 table makes the engine discontinuous, so an approximate match
  is not a match. An x86 path has to clear the same bar against the same
  reference.
- **The check that decides it** is `tests/run.sh`'s *SIMD backend matches the
  CPU baseline*, which runs `WASTE_BACKEND=cpu` against the dispatched path and
  compares logits. It has to run on an `--index-bits 6` container: the suite's
  default `WASTE_REF_MODEL` is a VQ3R conversion, and per `CLAUDE.md` running
  the suite on the wrong container shape is exactly how a load path once stayed
  broken through green runs.
- **No machine here can execute it.** Whoever writes it is also the only person
  who can test it, so the measurement matters more than the patch.

**Comment by marcobambini** [comment]
The doc sentence is qualified, in
[48b8afa](https://github.com/sqliteai/warp/commit/48b8afa) — kept and annotated
rather than rewritten, per that file's convention. The kernel itself is still
open and this issue stays as the tracking one.

The qualification says what you asked it to: the sentence is right about VQ3R
and §41 has the arithmetic behind it (256 bytes, sixteen vector registers of
thirty-two, no blocking makes it fit), and VQ4P breaks its premise rather than
its wording — 4x64 spends the same 24 bits per 8-weight vector, and 64 bytes
is four registers, which is exactly what `vqtbl4q_s8` addresses. Also in
there: the `vpermi2b` candidate, that `simd_avx512.c` is a slot to fill rather
than an `#ifdef`, both reasons it may not pay, and the acceptance bar —
bit-identity per §43, checked by *SIMD backend matches the CPU baseline* on an
`--index-bits 6` container.

## One correction, and it narrows this issue

> **VQ4P is the format `convert.py` produces by default**, so this is the path
> on every x86 build, not a corner case.

It is not the default. `tools/convert.py:863` has `--index-bits` at **8** and
`--stages` at **3**, so a default conversion is VQ3R — which is also what
`CLAUDE.md` says when it tells you to point `WASTE_REF_MODEL` at a default
conversion and calls `--index-bits 6` "a different kernel and a different
record fmt". VQ4P needs `--index-bits 6 --stages 4 --entries 64` asked for
explicitly, and `convert.py:903` refuses the flag without the other two.

So the scalar apply is not on every x86 build, and the urgency argument in the
issue does not stand as written.

**What replaces it is narrower and, I think, worse.** The shape that reaches
the scalar path is exactly the shape §50 left open as the 64-entry crossover.
So the population most likely to meet it is the population trying to answer
that question — and what they would measure is a scalar apply against a
vectorized one, then report it as a table-size result. A wrong number that
looks like an answer is more expensive than a slow path that announces itself,
particularly for a question this file has left open on purpose.

The rest of the issue is unaffected and stays as filed: §47's tuning table is
an ARM result presented as engine-wide guidance, the bit-identity bar, and the
requirement that whoever writes it is also the only person who can test it.

I also fixed the two small inaccuracies while verifying against `src/model.c`:
`VQ_TILE` is 64, not 16 — the four `vqtbl4q_s8` are per 16-row group, four
groups to a tile.

**Comment by marcobambini** [comment]
Closing this: it asked for two things, one is done and the other has moved to
its own issue with the scope corrected.

**The documentation half is done** —
[48b8afa](https://github.com/sqliteai/warp/commit/48b8afa). The sentence in
`docs/BACKENDS.md` is qualified rather than rewritten, per that file's
convention: it is right about VQ3R and §41 has the arithmetic behind it, and
VQ4P breaks its premise rather than its wording. §47's tuning table being an
ARM result is recorded there too, in the terms this issue asked for.

**The kernel half is #38.** It is scoped differently in one respect that
matters, and the correction came out of verifying this issue against the tree:
VQ4P is **not** what `convert.py` produces by default — `--index-bits` is 8
and `--stages` is 3, so a default conversion is VQ3R, and `--index-bits 6` is
refused without `--stages 4 --entries 64`. So the scalar apply is not on every
x86 build, and the urgency argument here does not stand as written.

What replaces it is narrower and, I think, worse, so it is the framing #38
leads with: the shape that reaches the scalar path is exactly the 64-entry
crossover §50 lists as not measured, so the population most likely to meet it
is the population trying to answer that question — and they would report an
ISA difference as a table-size result.

#38 also carries something this issue did not identify. The reason x86 has no
kernel is structural rather than incidental: `vq_rows_p6` is vectorized behind
an `#ifdef` in `model.c` and never moved behind `waste_kernels` the way
`mvq_rows_f32` and `lutb_range` did. So the first step is adding the dispatch
slot and moving the NEON body into it unchanged — bit-identical on ARM by
construction, reviewable **on this machine**, and the step where a mistake
would be cheapest to find. `WASTE_CPU_AVX512VBMI` also does not exist in the
feature enum yet, and `AVX512F && AVX512BW` is not sufficient to gate
`vpermi2b`.

Everything else survives unchanged and is restated there: the `vpermi2b`
candidate, the bit-identity bar per §43, *SIMD backend matches the CPU
baseline* as the deciding check, the requirement that it run on an
`--index-bits 6` container, both reasons it may not pay, and that whoever
writes it is the only person who can test it.
